### DISCUSSION TITLE
NGINX log to Seq

### DISCUSSION THREAD

**Discussion by fomcl** [datalust/seq-tickets#1922] [Help]
Hi,

New to Seq. I want to write nginx access logs to Seq using syslog (no Docker involved). I tried the Syslog.Input plugin (RFC 3164 or RFC 5424) and the GELF plugin. Can you provide a minimal nginx configuration with the recommended approach, which I believe is GELF?*)

I tried:
```
log_format graylog_json escape=json '{ "version": "1.1", "host": "$host", "short_message": "hi there, champ", "level": 5 }';
access_log syslog:server=myseqserver:12201,severity=debug graylog_json;
```

In the Seq events log I see "GELF processing failed". When I make a dump of the data using `tcpdump` I see GELF/json embedded in what appears to be a RFC3164 message.

An alternative to GELF is also fine, as long as I can query the various parts of the logging.

Thanks!
Albert-Jan

*) https://blog.datalust.co/seq-input-syslog/#method1windowsdockerinstallingseqinputsyslogdirectlyinseq

```
nginx version: nginx/1.20.1
Linux S0MGPC0003G.cbsp.nl 3.10.0-1160.95.1.el7.x86_64 #1 SMP Fri Jun 23 08:44:55 EDT 2023 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by nblumhardt** [comment]
Hi Albert-Jan, thanks for your message.

It sounds more like you'll need the _Seq.Input.Syslog_ plug-in than the GELF input for this; what kind of result do you get with it?

**Comment by fomcl** [comment]
Hi,

Thanks for helping me.

Given the following two formats:

```
    access_log syslog:server=s0aseq0009a.cbsp.nl:514,facility=local7,tag=nginx_rfc5424,severity=debug syslog_rfc5424;  # see appendix below
    access_log syslog:server=s0aseq0009a.cbsp.nl:514,facility=local7,tag=nginx_rfc3164,severity=debug;
```

... I see the following in SEQ:

![Capture](https://github.com/datalust/seq-tickets/assets/7590148/8322c7a1-1cba-413e-b268-bd238aaff9a6)

With neither format I am able to use/query the individual log variables, e.g. `$hostname`. The NGINX log is in SEQ, but it is not parsed.
The rfc5424 format has all the info that I want, e.g `$sent_http_x_username`.


Thank you!

Albert-Jan

Appendix
-----------
Here's the definition for the `syslog_rfc5424` custom nginx log format:
```
    log_format syslog_rfc5424 '<190>1'
        ' $time_iso8601'
        ' $hostname'
        ' nginx'
        ' $pid'
        ' $request_method'
        ' [nginx@32473'
        ' body_bytes_sent="$body_bytes_sent"'
        ' connection="$connection"'
        ' connection_requests="$connection_requests"'
        ' forwarded_for="$http_x_forwarded_for"'
        ' host="$host"'
        ' referer="$http_referer"'
        ' remote_addr="$remote_addr"'
        ' remote_user="$remote_user"'
        ' request="$request_uri"'
        ' request_time="$request_time"'
        ' status="$status"'
        ' time_iso8601="$time_iso8601"'
        ' time_local="$time_local"'
        ' user_agent="$http_user_agent"'
        ' sent_http_x_username="$sent_http_x_username"'
        ']';

```

**Reply by nblumhardt** [reply]
Thanks for following up. It looks as though perhaps the format reaching Seq over the wire isn't actually valid RFC5424, because the message has the literal text `"nginx_rfc5424: "` prepended at the start 🤔 - am I reading that correctly?

The syslog input's tests have a few examples of the format taken from the syslog spec that might be good to compare with what you're seeing from nginx, at: https://github.com/datalust/seq-input-syslog/blob/dev/squiflog/src/data/syslog.rs#L330

Writing to a local file and forwarding it with Fluent Bit is also fine, and worth doing if it will get you up and running faster, but RFC5424 should be accepted as-is by Seq, as long as it's spec-conformant.

If there's any raw data you have at hand that I can try pasting into the test suite, please do post it here, I'll be happy to give it a try! 👍

**Comment by fomcl** [comment]
Or would it be a better approach to write the nginx log locally (I am doing that already, as a csv file) and use `fluentd` or similar to tail/parse that and send it to SEQ? https://docs.fluentd.org/parser/csv. It seems so strange to me that a popular web server like NGINX is not supported out-of-the-box.

**Reply by fomcl** [reply]
> the message has the literal text "nginx_rfc5424: " prepended at the start 🤔 - am I reading that correctly?

Yes, I think this is the problem. It's actually prepended with [A] `nginx_rfc5424` (what is called "tag" in the nginx `access_log` directive) and [B] `: `. I can set the tag [A] to an empty string, but I think I have no control over [B]. Grrr. Will check again on Monday.

**Comment by fomcl** [comment]
Hi,

Given the drawback mentioned earlier I'm no longer using syslog. I came up with this and it does what I want (https still to do). I thought I'd leave this `fluentd` conf here for others. Any feedback is greatly appreciated. Thanks!

```
<source>
  @type tail
  @id input_tail
  tag my_access
  path /var/log/nginx/reverse-access_a.log
  pos_file /var/tmp/reverse-access_a.log.pos
  <parse>
    @type csv
    keys remote_addr,remote_user,time_iso8601,request,sent_http_x_username,status,body_bytes_sent,bytes_sent,http_referer,http_user_agent,http_x_forwarded_for,request_time
    delimiter ;
  </parse>
</source>

## http://clef-json.org/
<filter my_access**>
  @type record_transformer
  enable_ruby
  <record>
    "@t" "${record[\"time_iso8601\"]}"
    "@mt" "{sent_http_x_username} - {request} [{status}]"
  </record>
</filter>

## debug
#<match my_access**>
#  @type stdout
#</match>

## https://docs.fluentd.org/output/http
<match my_access**>
  @type http
  endpoint http://myserver.com:5341/api/events/raw?clef
  proxy none
  open_timeout 2
  http_method post
  content_type json
  json_array false
  headers {"X-Seq-ApiKey": "<<secret>>"}
  <format>
    @type json
  </format>
  <buffer>
    flush_interval 10s
  </buffer>
</match>
```

In the fluentd log I see
```
2023-08-24 15:55:47 +0200 [warn]: #0 fluent/log.rb:381:warn: '@' is the system reserved prefix. It works in the nested configuration for now but it will be rejected: @t
2023-08-24 15:55:47 +0200 [warn]: #0 fluent/log.rb:381:warn: '@' is the system reserved prefix. It works in the nested configuration for now but it will be rejected: @mt
```
which suggests I could improve the code (but it works!)

**Reply by nblumhardt** [reply]
Fantastic! Thanks for sharing your solution 👍
